bootstrap/kubeadm/controllers/kubeadmconfig_controller.go

```diff
@@ -144,25 +146,32 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(config, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
 	switch {
-	// Wait patiently for the infrastructure to be ready
+	// Wait for the infrastructure to be ready.
 	case !cluster.Status.InfrastructureReady:
 		log.Info("Infrastructure is not ready, waiting until ready.")
 		return ctrl.Result{}, nil
-	// bail super early if it's already ready
+	// Migrate plaintext data to secret.
+	case config.Status.BootstrapData != nil:
```

Comment:
Considering the lossiness of the conversion, maybe we should hold off on automated migration until whatever release is after v1alpha3?

Reply:
I'd like to keep the automatic migration, but I'll remove the line that sets the BootstrapData to `nil` here, that should allow to convert v1alpha3 to v1alpha2, unless it's a secret-only

Reply:
Then inside storeBootstrapData should we ignore the create error that secret may already exist?